File: SHT_example.c
/*
 * Copyright (c) 2010 Centre National de la Recherche Scientifique.
 * written by Nathanael Schaeffer (CNRS, LGIT, Grenoble, France).
 * 
 * nathanael.schaeffer@ujf-grenoble.fr
 * 
 * This software is governed by the CeCILL license under French law and
 * abiding by the rules of distribution of free software. You can use,
 * modify and/or redistribute the software under the terms of the CeCILL
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 * 
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL license and that you accept its terms.
 * 
 */

/** \example SHT_example.c
 \brief An example program that performs backward and forward Spherical Harmonic Transforms using SHTns.

  Compile using : \code make SHT_example \endcode
  \see \ref usage
**/

#include <stdio.h>
#include <stdlib.h>
#include <complex.h>
#include <math.h>
#include <fftw3.h>

#include <shtns.h>

/// a simple function that writes a vector to a file
void write_vect(char *fn, double *vec, int N);

/// a simple function that writes a matrix to a file.
void write_mx(char *fn, double *mx, int N1, int N2);


int main()
{
	long int lmax,mmax,nlat,nphi,mres, NLM;
	complex double *Slm, *Tlm;	// spherical harmonics l,m space
	double *Sh, *Th;		// real space : theta,phi
	long int i,im,lm;
	double t;

	lmax = 5;	nlat = 32;
	mmax = 3;	nphi = 10;
	mres = 1;
	NLM = shtns_init( sht_gauss, lmax, mmax, mres, nlat, nphi );
//	NLM = shtns_set_size(lmax, mmax, mres, sht_orthonormal | SHT_REAL_NORM);
//	shtns_precompute( sht_gauss, 0.0, nlat, nphi);

// Memory allocation : the use of fftw_malloc is required because we need proper 16-byte alignement.
// allocate spatial fields.
	Sh = (double *) fftw_malloc( NSPAT_ALLOC * sizeof(double));
	Th = (double *) fftw_malloc( NSPAT_ALLOC * sizeof(double));

// allocate SH representations.
	Slm = (complex double *) fftw_malloc( NLM * sizeof(complex double));
	Tlm = (complex double *) fftw_malloc( NLM * sizeof(complex double));

// SH_to_spat
	LM_LOOP( Slm[lm]=0.0;  Tlm[lm] = 0.0; )		/* this is the same as :
						for (lm=0;lm<NLM;lm++) {
							Slm[lm] = 0.0;	Tlm[lm] = 0.0;
						} */

//	Slm[LM(1,1)] = sh11_st();				// access to SH coefficient
	Slm[LM(2,0)] = 1.0;
// 	Slm[LiM(1,0)] = sh10_ct();
//	Slm[LiM(0,0)] = 0.5*sh00_1();
//	SH_to_spat(Slm,Sh);
	SHtor_to_spat_m0(Slm,Sh);
	write_vect("ylm",(double *) Slm,NLM*2);
	write_mx("spat",Sh,nphi,nlat);

// compute value of SH expansion at a given physical point.
 	double t2;
	SHqst_to_point(Tlm, Tlm, Slm, ct[nlat/3], 2.*M_PI/(mres*nphi),&t2,&t2,&t);
	printf("check if SH_to_point coincides with SH_to_spat : %f = %f\n",t,Sh[nlat/3]);
	printf("ct*st = %f\n",ct[nlat/3]*st[nlat/3]);

// check non-linear behaviour
	for (im=0;im<nphi;im++) {
		for (i=0;i<nlat;i++) {
			Sh[im*nlat+i] *= Sh[im*nlat+i];
		}
	}
	spat_to_SH(Sh,Tlm);
	write_vect("ylm_nl",(double *) Tlm,NLM*2);

// vector transform
	LM_LOOP( Slm[lm]=0.0;  Tlm[lm] = 0.0; )
	SHsphtor_to_spat(Slm,Tlm,Sh,Th);		// vector transform
	write_mx("spatt",Sh,nphi,nlat);
	write_mx("spatp",Th,nphi,nlat);

// spat_to_SH
	for (im=0;im<nphi;im++) {
		for (i=0;i<nlat;i++) {
			Sh[im*nlat+i] = ct[i];			// use cos(theta) array
		}
	}
	spat_to_SH(Sh,Slm);
	write_vect("ylm_v",(double *) Slm,NLM*2);
}


void write_vect(char *fn, double *vec, int N)
{
	FILE *fp;
	int i;
	
	fp = fopen(fn,"w");
	for (i=0;i<N;i++) {
		fprintf(fp,"%.6g ",vec[i]);
	}
	fclose(fp);
}

void write_mx(char *fn, double *mx, int N1, int N2)
{
	FILE *fp;
	int i,j;
	
	fp = fopen(fn,"w");
	for (i=0;i<N1;i++) {
		for(j=0;j<N2;j++) {
			fprintf(fp,"%.6g ",mx[i*N2+j]);
		}
		fprintf(fp,"\n");
	}
	fclose(fp);
}

File: SHT_example.f
c
c  Copyright (c) 2010-2011 Centre National de la Recherche Scientifique.
c  written by Nathanael Schaeffer (CNRS, ISTerre, Grenoble, France).
c  
c  nathanael.schaeffer@ujf-grenoble.fr
c  
c  This software is governed by the CeCILL license under French law and
c  abiding by the rules of distribution of free software. You can use,
c  modify and/or redistribute the software under the terms of the CeCILL
c  license as circulated by CEA, CNRS and INRIA at the following URL
c  "http://www.cecill.info".
c  
c  The fact that you are presently reading this means that you have had
c  knowledge of the CeCILL license and that you accept its terms.
c  


* A Fortran example program that performs backward and forward Spherical Harmonic Transforms using SHTns
      PROGRAM SHT_example

      IMPLICIT NONE
c import useful parameters for shtns initialization
      include 'shtns.f'

      integer*4 lmax, mmax, mres
      integer*4 nlat, nphi
      integer*4 nlm
      integer*4 layout
      integer*4 norm
      real*8 eps_polar
      complex*16, allocatable :: Slm(:), Tlm(:)
      real*8, allocatable :: Sh(:,:), Th(:,:)

      integer i,lm, m

c set size of transform
      lmax = 5
      mmax = 2
      mres = 2
      nphi = 8
      nlat = 6

c compute sizes required for arrays.
      call shtns_calc_nlm(nlm, lmax, mmax, mres)
      print*,'NLM=',nlm

c init SHT. SHT_PHI_CONTIGUOUS is defined in 'shtns.f'
      layout = SHT_PHI_CONTIGUOUS
      call shtns_init_sh_gauss(layout, lmax, mmax, mres, nlat, nphi)

c alternatively, you can use the two following calls, giving more control on the initialization process,
c namely you can choose a normalization with 'norm' and control the polar optimization
c with 'eps_polar' : from 0 (no optimization) to 1e-6 (agressive optimization).
!      norm = SHT_ORTHONORMAL + SHT_REAL_NORM
!      call shtns_set_size_(lmax, mmax, mres, norm)
!      eps_polar = 1.e-10
!      call shtns_precompute_(SHT_GAUSS, layout, eps_polar, nlat, nphi)

c allocate memory for spectral and spatial representation
      allocate ( Slm(1:nlm), Tlm(1:nlm) )
      allocate ( Sh(1:nphi,1:nlat), Th(1:nphi,1:nlat) )
      Slm(:) = 0.0

c get index for given l and m
      call shtns_lmidx(lm, 1, 0)
      Slm(lm) = 1.0
      print*
      print*,Slm(:)

      call shtns_sh_to_spat(Slm, Sh)

c print all spatial data
      print*
      do i=1,nphi
        print*,'phi index=',i
        print*,Sh(i,:)
      enddo

      call shtns_spat_to_sh(Sh, Slm)

      print*
      print*,Slm(:)

c print SH data, grouping by m  (shows how to access SH coefficients)
      print*
      do m=0, mres*mmax, mres
        print*,'m=', m
        call shtns_lmidx(lm,m,m)
        print*,Slm(lm : lm+(lmax-m))
      enddo

      stop
      END
